cmd/containerd-shim-runc-v2/process/io.go

```diff
@@ -307,6 +329,180 @@ func NewBinaryIO(ctx context.Context, id string, uri *url.URL) (_ runc.IO, err e
 	}, nil
 }
 
+// NewAttachableBinary runs a custom binary process and opens attachable fifos for pluggable shim logging
+func NewAttachableBinary(ctx context.Context, id string, uri *url.URL) (_ *cio.FIFOSet, err error) {
+	binaryFifos, err := cio.NewFIFOSetInDir(defaults.DefaultFIFODir, id, true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create birany fifos: %w", err)
+	}
+
+	binaryFifosPipes, err := openBinaryFifos(ctx, binaryFifos)
+	if err != nil {
+		return nil, err
+	}
+
+	attachableFifos, err := cio.NewFIFOSetInDir(filepath.Join(defaults.DefaultFIFODir, "attach"), id, true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create attachable fifos: %w", err)
+	}
+	type closer func() error
+	var closers []closer
+
+	defer func() {
+		if err == nil {
+			return
+		}
+		result := []error{err}
+		for _, fn := range closers {
+			result = append(result, fn())
+		}
+		err = errors.Join(result...)
+	}()
+
+	pipesAttachableFifosPipes, err := openAttachableFifos(ctx, attachableFifos)
```

Comment:
Will it be problematic if there are no consumers of these FIFOs? Pipe has limited capacity and when it's full the writing side may block or fail.

Reply:
by default pipe would block when it's full, but with [O_NONBLOCK](https://github.com/containerd/containerd/pull/10020/files#diff-6cc4a8120115484ccbffe4499ddff51097c92da6bfe659905878c03234bc266bR489) flag coming data will be lost and new data will be queued to the writer, so attaching tails logs and this is the intended behavior

Reply:
according to [pipe(7)](https://man7.org/linux/man-pages/man7/pipe.7.html):

> A pipe has a limited capacity.  If the pipe is full, then a
       [write(2)](https://man7.org/linux/man-pages/man2/write.2.html) will block or fail, depending on whether the O_NONBLOCK
       flag is set (see below).

IIRC non-block flag will simply result in write failure rather than blocking the caller. Either way the result is not favorable.

Not sure if i misunderstood this :)